### PR TITLE
[Fix #742] Id, key and name refactor

### LINKED ISSUE #742: workflow.id and workflow.key is duplicate
**What would you like to be added**:

Only keep one of id or key property in workflow.

**Why is this needed**:

As the description:

```
The id property defines the unique, domain-specific workflow identifier, for example "orders", "payment", etc.

The key property defines the unique, domain-specific workflow identifier. It can be used when the id property is auto-generated by a content-management system for example. In these cases, you can specify the key property to be the domain-specific identifier of the workflow definition. The id and key properties are mutually exclusive, meaning you cannot define both.
```

The id & key property is exclusive，we can only have one of them, but when id is auto-generated (which means it have an value), we cannot define the key property, so the use case is wrong.

### PR DESCRIPTION
**Pull request by fjtirado**
Id is removed.
Name remains mandatory and is clarified that should be unique. Key description is updated and becomes optional in all cases. Version is assumed to be latest if not specified

**Many thanks for submitting your Pull Request :heart:!**

**Please specify parts of this PR update:**

- [X] Specification
- [X] Schema
- [X] Examples
- [ ] Extensions
- [ ] Roadmap
- [ ] Use Cases
- [ ] Community
- [ ] TCK
- [ ] Other

**Discussion or Issue link**:

Fixes https://github.com/serverlessworkflow/specification/issues/742

**What this PR does / why we need it**:

**Special notes for reviewers**:

**Additional information:**

### CODE REVIEW

File: specification.md [hunk 1]
```diff
@@ -1925,11 +1925,10 @@ definition "id" must be a constant value.
...
 | Parameter | Description | Type | Required |
 | --- | --- |  --- | --- |
-| id | Workflow unique identifier | string | yes (if `key` is not defined) |
-| key | Domain-specific workflow identifier | string | yes (if `id` is not defined) |
-| name | Workflow name | string | no |
+| name | Workflow definition unique identifier (which must be human readable) | string | yes |
+| version | Workflow version. MUST respect the [semantic versioning](https://semver.org/) format. If not provided, latest is assumed | string | no |
 | description | Workflow description | string | no |
-| version | Workflow version. MUST respect the [semantic versioning](https://semver.org/) format | string | no |
+| key | Expression that will be used to generate a domain-specific workflow instance identifier  | string | no |
```
Comment (fjtirado) [inline specification.md:1931]: I left key name as it is. In an ideal world, probably the name should be something like instance-key-generator, but to generate another backward incompatibility because of that seems not pratical to me, so I vote for the name "key" to stand

**Review by ricardozanini — COMMENTED** [review]
Looks good! But there are a few snippets throughout the `specification.md` that need to be removed too. Can you search for `id:` or `"id":`

**Review by ricardozanini — APPROVED** [review]

**Comment by fjtirado** [comment]
@cdavernas Can you please take a look?

File: schema/auth.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "$id": "https://serverlessworkflow.io/schemas/0.8/auth.json",
+  "$id": "https://serverlessworkflow.io/schemas/0.9/auth.json",
```
Comment (cdavernas) [inline schema/auth.json:2]: Shouldn't version be updated at release time only, as referenced uri does not yet exist?

@ricardozanini @fjtirado @JBBianchi

File: schema/workflow.json [hunk 3]
```diff
@@ -1,33 +1,26 @@
...
-    "id": {
-      "type": "string",
-      "description": "Workflow unique identifier",
-      "minLength": 1
-    },
-    "key": {
+    "name": {
       "type": "string",
-      "description": "Domain-specific workflow identifier",
+      "description": "Workflow definition unique identifier (which must be human readable)",
```
Comment (cdavernas) [inline schema/workflow.json:9]: Shouldn't the concatenation of both `name` and `version` be unique, instead of just the `name`?

If so, what about something like:

```
The name that identifies the workflow definition, and which, when combined with its version, forms a unique identifier. 
```

File: schema/workflow.json [hunk 4]
```diff
@@ -1,33 +1,26 @@
...
       "type": "string",
       "description": "Workflow description"
     },
-    "version": {
+    "key": {
       "type": "string",
-      "description": "Workflow version",
-      "minLength": 1,
-      "pattern": "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$"
+      "description": "Expression that will be used to generate a domain-specific workflow instance identifier"
```
Comment (cdavernas) [inline schema/workflow.json:23]: I would rephrase to something like the following, to include the optional aspect of the property:

```
Optional expression used to generate a domain-specific workflow instance identifier
```

File: schema/workflow.json [hunk 5]
```diff
@@ -1,33 +1,26 @@
...
   "description": "Serverless Workflow specification - workflow schema",
   "type": "object",
   "properties": {
-    "id": {
-      "type": "string",
-      "description": "Workflow unique identifier",
-      "minLength": 1
-    },
-    "key": {
+    "name": {
```
Comment (cdavernas) [inline schema/workflow.json:7]: Shouldn't the definition's name follow workflow [naming conventions](https://github.com/serverlessworkflow/specification/blob/main/specification.md#naming-convention)?

> Specifically, names must be lowercase, start and end with an alphanumeric character, and consist entirely of alphanumeric characters with optional isolated medial dashes '-' (i.e., dashes must not be adjacent to each other).

I personally would advocate enforcing it, especially now that it's to become (part of) an identifier, and would make the spec more consistent (as we enforce it for all other `name` properties).

If so, there's the need to add a pattern property, and possibly update the description.

**Review by cdavernas — CHANGES_REQUESTED** [review]
Thank you so much for you work! ❤️ 

I've got some comments, though only a single one is a pita 😄

File: schema/workflow.json (hunk 3 above)
Comment (fjtirado) [reply schema/workflow.json:9]: I think you are overcomplicating it ;)
The name should be unique regardless whatever is in the version, so combined with the version it will allways be unique.

File: schema/workflow.json (hunk 4 above)
Comment (fjtirado) [reply schema/workflow.json:23]: I think the fact that key is optional is indicated in other place and feel it to be redundant while writing it, but it does not harm to emphasize it

File: schema/workflow.json (hunk 5 above)
Comment (fjtirado) [reply schema/workflow.json:7]: ok, to be honest I did not realize the convention was not being enforced because of the lack of regular expression in the schema

File: schema/auth.json (hunk 2 above)
Comment (fjtirado) [reply schema/auth.json:2]: hmmm, I found confusing keeping the 0.8 there, but as you prefer

Comment (cdavernas) [reply schema/auth.json:2]: Just asking. I saw that @ricardozanini requested it in my PR, too, so let's go with it.

File: schema/workflow.json (hunk 3 above)
Comment (cdavernas) [reply schema/workflow.json:9]: The problem I have is with the `unique` adjective you are using. 

Given the terminology used, I definitly see how one could then confuse the `name` with the concept of an `id`: only one 'my-cool-workflow' can ever exist.

In my opinion, we should be as explicit as possible, especially in the `specification.md` file, so that authors and implementers understand that, to reference a workflow definition in a UNIQUE fashion, you should instead use `{name}:{version}`, for example. In other words, you can have multiple 'my-cool-workflow' definitions, but they must all have different versions.

File: schema/workflow.json (hunk 5 above)
Comment (cdavernas) [reply schema/workflow.json:7]: Yeah, no worries. I think it was voluntarily overlooked precisly because we though of addressing it in this PR 😆

File: schema/workflow.json (hunk 3 above)
Comment (fjtirado) [reply schema/workflow.json:9]: Ok, the only problem I find with your re phrasing is that version is optional, and somehow someone can understood that is not.

Comment (cdavernas) [reply schema/workflow.json:9]: Yeah, sorry about that, that was something we discussed yesterday in the weekly.

What was proposed is to provide `version` a default `latest` value, so that even if it's optional and not set by the user, a value is always supplied, making `version` non-nullable.

File: specification.md [hunk 6]
```diff
@@ -643,8 +643,8 @@ a workflow with a single event state and show how data filters can be combined.
 
 ```json
 {
-    "id": "GreetCustomersWorkflow",
-    "name": "Greet Customers when they arrive",
+    "name": "GreetCustomersWorkflow",
```
Comment (cdavernas) [inline specification.md:646]: Could you please fix example names so that they follow the naming convention?

File: specification.md [hunk 7]
```diff
@@ -1910,26 +1910,25 @@ we can use this expression in the workflow "version" parameter:
...
 
 ### Workflow Definition Structure
 
 | Parameter | Description | Type | Required |
 | --- | --- |  --- | --- |
-| id | Workflow unique identifier | string | yes (if `key` is not defined) |
-| key | Domain-specific workflow identifier | string | yes (if `id` is not defined) |
-| name | Workflow name | string | no |
+| name | The name that identifies the workflow definition, and which, when combined with its version, forms a unique identifier. | string | yes |
+| version | Workflow version. MUST respect the [semantic versioning](https://semver.org/) format. If not provided, latest is assumed | string | no |
```
Comment (cdavernas) [inline specification.md:1929]: Could you put latest inside a ` MD block, so that it looks like this: `latest`. 

It is imho more readable, and allows users to quickly identify "tagged" values.

**Review by cdavernas — CHANGES_REQUESTED** [review]
Thank you for your work, once again!

I'm a pita, sorry man: I left a couple of small comments.

Also, could you be so kind to update the `roadmap.md` file to include changes of this PR?

File: schema/workflow.json [hunk 8]
```diff
@@ -1,33 +1,27 @@
...
+	  "pattern": "^[a-z0-9](-?[a-z0-9])*$",
       "minLength": 1
     },
-    "name": {
+    "version": {
       "type": "string",
-      "description": "Workflow name",
-      "minLength": 1
+      "description": "Workflow version",
+      "pattern": "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$"
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply schema/workflow.json:16]: I any case, although related, this version discussion is outside the issue scope.

File: specification.md [hunk 9]
```diff
@@ -1910,26 +1910,25 @@ we can use this expression in the workflow "version" parameter:
 
 ```json
 {
-   "id": "MySampleWorkflow",
-   "name": "Sample Workflow",
+   "name": "MySampleWorkflow",
```
Comment (cdavernas) [inline specification.md:1913]: Name is invalid here, too

File: specification.md [hunk 10]
```diff
@@ -1963,10 +1962,9 @@ definition "id" must be a constant value.
 
 ```json
 {
-   "id": "sampleWorkflow",
+   "name": "sampleWorkflow",
```
Comment (cdavernas) [inline specification.md:1965]: Invalid name

File: specification.md [hunk 11]
```diff
@@ -1981,10 +1979,9 @@ definition "id" must be a constant value.
 <td valign="top">
 
 ```yaml
-id: sampleWorkflow
+name: sampleWorkflow
```
Comment (cdavernas) [inline specification.md:1982]: Invalid name

File: specification.md [hunk 12]
```diff
@@ -2169,10 +2161,9 @@ Here is an example of using external resource for function definitions:
 
 ```json
 {
-   "id": "sampleWorkflow",
+   "name": "sampleWorkflow",
```
Comment (cdavernas) [inline specification.md:2164]: Invalid

File: specification.md [hunk 13]
```diff
@@ -2206,10 +2197,9 @@ Here is an example of using external resource for event definitions:
...
```
Comment (cdavernas) [inline specification.md:2200]: Invalid

File: specification.md [hunk 14]
```diff
@@ -3187,7 +3176,6 @@ and our workflow is defined as:
 <td valign="top">
 
 ```yaml
-id: sendConfirmWorkflow
 name: SendConfirmationForCompletedOrders
```
Comment (cdavernas) [inline specification.md:3179]: Invalid

File: specification.md [hunk 15]
```diff
@@ -5600,7 +5588,7 @@ To give an example, let's say that in our workflow definition we define the time
 
 ```json
 {
-   "id": "testWorkflow",
+   "name": "testWorkflow",
```
Comment (cdavernas) [inline specification.md:5591]: Invalid

File: specification.md [hunk 16]
```diff
@@ -6316,8 +6304,8 @@ Here is an example of metadata attached to the core workflow definition:
 
 ```json
 {
-  "id": "processSalesOrders",
-  "name": "Process Sales Orders",
+  "name": "processSalesOrders",
```
Comment (cdavernas) [inline specification.md:6307]: Invalid

File: specification.md [hunk 17]
```diff
@@ -6352,8 +6340,8 @@ Implementations may use this keyword to give access to any relevant information
 ```json
 
 {
-  "id": "processSalesOrders",
-  "name": "Process Sales Orders",
+  "name": "processSalesOrders",
```
Comment (cdavernas) [inline specification.md:6343]: Invalid

**Review by cdavernas — CHANGES_REQUESTED** [review]

File: specification.md (hunk 9 above)
Comment (fjtirado) [reply specification.md:1913]: Yes, Im currently searching for mislaligments on specificatin.md

**Comment by fjtirado** [comment]
@cdavernas I think I fixed all names misalignment on specification.md (hopefully ;))

File: specification.md [hunk 18]
```diff
@@ -4201,19 +4189,19 @@ should not fail workflow execution.
...
 
-If `string` type, it defines the unique id of the sub-workflow to be invoked.
-This short-hand definition can be used if sub-workflow lookup is done only by its `id`
+If `string` type, it defines the unique name of the sub-workflow to be invoked.
+This short-hand definition can be used if sub-workflow lookup is done only by its `name`
 property and not its `version` property.
 
 ```json
-"subFlowRef": "mySubFlowId"
+"subFlowRef": "mySubFlowName"
```
Comment (JBBianchi) [inline specification.md:4197]: Invalid name, should be "my-sub-flow-name".

File: specification.md [hunk 19]
```diff
@@ -6110,23 +6097,23 @@ likely to be forced to stop before reaching its completion. This can result in u
...
 The end definitions `continueAs` can be either of type `string` or `object`.
-If string type, it contains the unique workflow id of the workflow that the execution should continue as, for example:
+If string type, it contains the unique workflow name of the workflow that the execution should continue as, for example:
 
 
 ```json
 { 
   "end": {
-    "continueAs": "myworkflowid"
+    "continueAs": "myworkflowname"
```
Comment (JBBianchi) [inline specification.md:6110]: (Optional) Improve name: my-workflow-name

File: specification.md [hunk 20]
```diff
@@ -6110,23 +6097,23 @@ likely to be forced to stop before reaching its completion. This can result in u
...
   }
 }
 ```
 
 Defining this should stop the current workflow execution, and continue execution as a new workflow instance of the 
-workflow which defines the workflow id of "myworkflowid". The state data where this is define should 
+workflow which defines the workflow name of "myworkflowbane". The state data where this is define should
```
Comment (JBBianchi) [inline specification.md:6116]: (Optional) Improve name: my-workflow-name

File: examples/accumulate-room-readings.json [hunk 21]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "roomreadings",
     "name": "roomreadings",
```
Comment (JBBianchi) [inline examples/accumulate-room-readings.json:2]: (Optional) Improve name: room-readings

File: examples/book-lending.json [hunk 22]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "booklending",
     "name": "booklending",
```
Comment (JBBianchi) [inline examples/book-lending.json:2]: (Optional) Improve name: book-lending

File: examples/car-vitals-checks-subflow.json [hunk 23]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "vitalscheck",
     "name": "vitalscheck",
```
Comment (JBBianchi) [inline examples/car-vitals-checks-subflow.json:2]: (Optional) Improve name: vitals-check

File: examples/car-vitals-checks.json [hunk 24]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "checkcarvitals",
     "name": "checkcarvitals",
```
Comment (JBBianchi) [inline examples/car-vitals-checks.json:2]: (Optional) Improve name: check-car-vitals

File: examples/continuing-as-a-new-execution.json [hunk 25]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "notifycustomerworkflow",
     "name": "notifycustomerworkflow",
```
Comment (JBBianchi) [inline examples/continuing-as-a-new-execution.json:2]: (Optional) Improve name: notify-customer-workflow

File: examples/filling-a-glass-of-water.json [hunk 26]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "fillglassofwater",
     "name": "fillglassofwater",
```
Comment (JBBianchi) [inline examples/filling-a-glass-of-water.json:2]: (Optional) Improve name: fill-glass-of-water

File: examples/new-patient-onboarding.json [hunk 27]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "patientonboarding",
     "name": "patientonboarding",
```
Comment (JBBianchi) [inline examples/new-patient-onboarding.json:2]: (Optional) Improve name: patient-on-boarding

File: examples/perform-customer-credit-check.json [hunk 28]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "customercreditcheck",
     "version": "1.0.0",
     "specVersion": "0.8",
     "name": "customercreditcheck",
```
Comment (JBBianchi) [inline examples/perform-customer-credit-check.json:4]: (Optional) Improve name: customer-credit-check

File: examples/process-transactions.json [hunk 29]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "customerbankingtransactions",
     "name": "customerbankingtransactions",
```
Comment (JBBianchi) [inline examples/process-transactions.json:2]: (Optional) Improve name: customer-banking-tansactions

File: examples/reusing-function-and-event-definitions.json [hunk 30]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "paymentconfirmation",
     "version": "1.0.0",
     "specVersion": "0.8",
     "name": "paymentconfirmation",
```
Comment (JBBianchi) [inline examples/reusing-function-and-event-definitions.json:4]: (Optional) Improve name: payment-confirmation

File: examples/send-cloudevent-on-workflow-completion.json [hunk 31]
```diff
@@ -1,5 +1,4 @@
 {
-    "id": "sendcloudeventonprovision",
     "version": "1.0.0",
     "specVersion": "0.8",
     "name": "sendcloudeventonprovision",
```
Comment (JBBianchi) [inline examples/send-cloudevent-on-workflow-completion.json:4]: (Optional) Improve name: send-cloud-event-on-provision

**Review by JBBianchi — APPROVED** [review]
Great work @fjtirado ! Thanks a lot for your contribution.

**Comment by ricardozanini** [comment]
@cdavernas wanna take a second look?

**Review by cdavernas — APPROVED** [review]
LGTM! Many thanks @fjtirado!

**Comment by cdavernas** [comment]
@ricardozanini Feel free to merge when ready: it has been approved by majority, and has been out for a month.
